lib/trie/trie.go

```diff
@@ -80,52 +82,50 @@ func (t *Trie) handleTrackedDeltas(success bool, pendingDeletedMerkleValues map[
 	}
 }
 
-func (t *Trie) prepLeafForMutation(currentLeaf *Node,
+func (t *Trie) prepForMutation(currentNode *Node,
 	copySettings node.CopySettings,
-	pendingDeletedMerkleValues map[string]struct{}) (newLeaf *Node) {
-	if currentLeaf.Generation == t.generation {
-		// no need to deep copy and update generation
-		// of current leaf.
-		newLeaf = currentLeaf
-	} else {
-		newLeaf = updateGeneration(currentLeaf, t.generation, pendingDeletedMerkleValues, copySettings)
+	pendingDeletedMerkleValues map[string]struct{}) (
+	newNode *Node, err error) {
+	isRoot := currentNode == t.root
+	err = registerDeletedMerkleValue(currentNode, isRoot,
+		pendingDeletedMerkleValues)
+	if err != nil {
+		return nil, fmt.Errorf("registering deleted node: %w", err)
 	}
-	newLeaf.SetDirty()
-	return newLeaf
-}
 
-func (t *Trie) prepBranchForMutation(currentBranch *Node,
-	copySettings node.CopySettings,
-	pendingDeletedMerkleValues map[string]struct{}) (newBranch *Node) {
-	if currentBranch.Generation == t.generation {
+	if currentNode.Generation == t.generation {
 		// no need to deep copy and update generation
-		// of current branch.
-		newBranch = currentBranch
+		// of current node.
+		newNode = currentNode
 	} else {
-		newBranch = updateGeneration(currentBranch, t.generation, pendingDeletedMerkleValues, copySettings)
+		newNode = currentNode.Copy(copySettings)
+		newNode.Generation = t.generation
 	}
-	newBranch.SetDirty()
-	return newBranch
+	newNode.SetDirty()
+	return newNode, nil
 }
 
-// updateGeneration is called when the currentNode is from
-// an older trie generation (snapshot) so we deep copy the
-// node and update the generation on the newer copy.
-func updateGeneration(currentNode *Node, trieGeneration uint64,
-	deletedMerkleValues map[string]struct{}, copySettings node.CopySettings) (
-	newNode *Node) {
-	newNode = currentNode.Copy(copySettings)
-	newNode.Generation = trieGeneration
+func registerDeletedMerkleValue(node *Node, isRoot bool,
+	pendingDeletedMerkleValues map[string]struct{}) (err error) {
```

Comment:
consider defining a type for `map[string]struct{}`? maybe something like `type mappedMerklesValues map[string]struct{}`

Reply:
I would rather have it without a custom type (for clarity/explicitness/portability) if that's ok with you. Otherwise, I can do this in a future PR (since deletedMerkleValues `map[string]struct{}` are already in development).

Reply:
Never mind, stuff is getting out of hands on #2881 and I need a custom struct with methods in the end... So good suggestion :+1:

Reply:
Done in #2896 (next branch) to avoid adding even more deltas in this branch which is already thick.
I will definitely use that in the next-next branch to track inserted node hashes in the trie.